crates/wasi-common/src/sys/unix/bsd/hostcalls_impl.rs

```diff
@@ -296,3 +203,85 @@ pub(crate) fn fd_advise(
 
     Ok(())
 }
+
+pub(crate) fn fd_readdir<'a>(
+    os_file: &'a mut OsFile,
+    cookie: wasi::__wasi_dircookie_t,
+) -> Result<impl Iterator<Item = Result<Dirent>> + 'a> {
+    use std::sync::Mutex;
+
+    let dir = match os_file.dir {
+        Some(ref mut dir) => dir,
+        None => {
+            // We need to duplicate the fd, because `opendir(3)`:
+            //     Upon successful return from fdopendir(), the file descriptor is under
+            //     control of the system, and if any attempt is made to close the file
+            //     descriptor, or to modify the state of the associated description other
+            //     than by means of closedir(), readdir(), readdir_r(), or rewinddir(),
+            //     the behaviour is undefined.
+            let fd = (*os_file).try_clone()?;
+            let dir = Dir::from(fd)?;
+            os_file.dir.get_or_insert(Mutex::new(dir))
+        }
+    };
+
+    // Seek if needed. Unless cookie is wasi::__WASI_DIRCOOKIE_START,
+    // new items may not be returned to the caller.
+    if cookie == wasi::__WASI_DIRCOOKIE_START {
+        log::trace!("     | fd_readdir: doing rewinddir");
+        dir.lock().unwrap().rewind();
+    } else {
+        log::trace!("     | fd_readdir: doing seekdir to {}", cookie);
+        let loc = unsafe { SeekLoc::from_raw(cookie as i64) };
+        dir.lock().unwrap().seek(loc);
+    }
+
+    Ok(DirIter(dir.lock().unwrap()).map(|entry| {
+        let (entry, loc): (Entry, SeekLoc) = entry?;
+        Ok(Dirent {
+            name: entry
+                // TODO can we reuse path_from_host for CStr?
+                .file_name()
+                .to_str()?
+                .to_owned(),
+            ino: entry.ino(),
+            ftype: entry.file_type().into(),
+            // Set cookie manually:
+            // * on macOS d_seekoff is not set for some reason
+            // * on FreeBSD d_seekoff doesn't exist; there is d_off but it is
+            //   not equivalent to the value read from telldir call
+            cookie: loc.to_raw().try_into()?,
+        })
+    }))
+}
+
+struct DirIter<'a>(MutexGuard<'a, Dir>);
```

Comment:
Could we possibly have a common `DirIter` implementation for all *nixes? 

On Linux, nothing will prevent the caller to execute concurrent `fd_readdir`'s on the same directory stream, so I don't see why we should prevent it on BSD. In particular `readdir(2)` on Linux says: (emphasis added)
> In cases where multiple threads must read from the same directory stream, _using readdir() with external synchronization is still preferable_ to the use of the deprecated readdir_r(3) function

This means, that code which doesn't have do any synchronization itself for `readdir`ing a single directory stram from multiple threads will be invalid anyway, so we can probably get away with not having mutexes here. Unfortunately, this would probably require adding some `unsafe` code, but we might want to do it for performance reasons.

In particular, when threads are introduced to WASI, thread-safety of various function will have to be evaluated.

Reply:
I think I'd prefer to leave it behind a `Mutex` as-is for now. This change was not introduced in this PR, but only refactored out from the actual implementation of `fd_readdir` on BSD that we've had until now. I'm OK with some code duplication between Linux and BSD, especially since on BSD we still need to recoved the current location using `telldir` as it's not available to be read directly from the `dirent` entry.

Reply:
Then we could possibly have a system-dependent implementation of `Entry` which would reuse the `d_off` value on Linux and call `telldir` on other nixes, or just call telldir everywhere (it's just a glibc call, so while it's not free, it's cheap).

But I'm fine with addressing this in a subsequent PR.

Reply:
Agreed to address this in a subsequent PR. :-)